app/src/main/java/org/dicekeys/app/fragments/AssembleFragment.kt

```diff
@@ -1,19 +1,220 @@
-import org.dicekeys.app.viewmodels.DiceKeyViewModel
+import org.dicekeys.app.databinding.*
+import org.dicekeys.app.extensions.clearNavigationResult
+import org.dicekeys.app.extensions.getNavigationResult
+import org.dicekeys.app.fragments.backup.BackupFragment
+import org.dicekeys.app.repositories.DiceKeyRepository
+import org.dicekeys.app.viewmodels.AssembleViewModel
+import org.dicekeys.app.views.DiceKeyView
+import org.dicekeys.app.views.StickerTargetSheetView
+import org.dicekeys.dicekey.DiceKey
+import org.dicekeys.dicekey.Face
+import org.dicekeys.dicekey.FaceRead
+import javax.inject.Inject
 
 @AndroidEntryPoint
-class AssembleFragment: AppFragment<AssembleFragmentBinding>(R.layout.assemble_fragment) {
+class AssembleFragment : AppFragment<AssembleFragmentBinding>(R.layout.assemble_fragment) {
 
+    @Inject
+    lateinit var diceKeyRepository: DiceKeyRepository
+
+    private lateinit var pagerAdapter: AssemblePagerAdapter
+
+
+    val viewModel: AssembleViewModel by viewModels()
 
     override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
         super.onViewCreated(view, savedInstanceState)
 
+        binding.vm = viewModel
+
+        getNavigationResult<String>(ScanFragment.READ_DICEKEY)?.observe(viewLifecycleOwner) {
+            it?.let {
+                clearNavigationResult(ScanFragment.READ_DICEKEY)
+
+                FaceRead.diceKeyFromJsonFacesRead(it)?.let { faceRead ->
+
+                    DiceKey(faces = faceRead.faces.map {
+                        Face(letter = it.letter, digit = it.digit, orientationAsLowercaseLetterTrbl = it.orientationAsLowercaseLetterTrbl)
+                    }).also {
+                        viewModel.setDiceKey(it)
+                    }
+                }
+            }
+        }
+
+        getNavigationResult<Boolean>(BackupFragment.VALID_BACKUP)?.observe(viewLifecycleOwner) {
+            if (it) {
+                viewModel.diceKeyBackedUp.postValue(true)
+            }
+        }
+
+        pagerAdapter = AssemblePagerAdapter(childFragmentManager)
+
+        binding.viewPager.also {
+            it.adapter = pagerAdapter
+            it.addOnPageChangeListener(object : ViewPager.OnPageChangeListener {
+                override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {}
+
+                override fun onPageSelected(position: Int) {
+                    // viewModel.setPage(position)
+                }
+
+                override fun onPageScrollStateChanged(state: Int) {}
+
+            })
+            // Prevent swipe gestures
+            it.setOnTouchListener { _, _ -> true }
+        }
+
+        binding.progressBar.max = pagerAdapter.count - 1
+
+        binding.btnNext.setOnClickListener {
+            if (viewModel.page.value == pagerAdapter.count - 1) {
+                val diceKey = viewModel.diceKey.value
+                if (diceKey == null) {
+                    findNavController().popBackStack()
```

Comment:
If the DiceKey somehow becomes null, this forces us back up?

If this is a common check, we could do have a function that getDiceKeyOrEscapeViewIfAbsent() which will navigate us out if the diceKey field is null, then wrap the below else clause instead in something like...

`getDiceKeyOrEscapeViewIfAbsent()? let { diceKey ->`